vllm/v1/core/scheduler.py

```diff
@@ -386,18 +402,21 @@ def _make_cached_request_data(
         num_regular_tokens = num_scheduled_tokens - num_scheduled_spec_tokens
         new_token_ids = request.all_token_ids[
             num_computed_tokens:num_computed_tokens + num_regular_tokens]
-        req_data = self._cached_reqs_data.get(request.request_id)
-        if req_data is not None:
+
+        req_data_queue = self._cached_reqs_data.get(request.request_id)
+        if req_data_queue:
+            req_data = req_data_queue.get()
             req_data.resumed_from_preemption = resumed_from_preemption
             req_data.new_token_ids = new_token_ids
             req_data.new_block_ids = new_block_ids
             req_data.num_computed_tokens = num_computed_tokens
         else:
+            # No cached request data, or all cached request data has been
+            # used by the scheduled requests.
             req_data = CachedRequestData.from_request(request,
                                                       resumed_from_preemption,
                                                       new_token_ids,
                                                       new_block_ids)
-            self._cached_reqs_data[request.request_id] = req_data
```

Comment:
Don't we need `req_data_queue.put_no_wait(req_data)` here?

Reply:
No. The current logic is that every `scheduler_output` holds the cached request data objects, and these "on the fly" objects are not in `self._cached_reqs_data` so that they won't be reused by another batch before the current batch is finished. We put cached request data objects back to `self._cached_reqs_data` in `update_from_output`.